File: .github/scripts/validate_chart_release_versions.py
#!/usr/bin/env python3
"""Validate that changed charts declare unreleased chart versions."""

from __future__ import annotations

import argparse
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
CHARTS_DIR = ROOT / "charts"


def run(command: list[str], *, cwd: Path = ROOT) -> str:
    completed = subprocess.run(
        command,
        cwd=cwd,
        check=False,
        capture_output=True,
        text=True,
    )
    if completed.returncode != 0:
        if completed.stdout:
            print(f"stdout:\n{completed.stdout}", file=sys.stderr)
        if completed.stderr:
            print(f"stderr:\n{completed.stderr}", file=sys.stderr)
        raise RuntimeError(f"{' '.join(command)} failed with exit code {completed.returncode}")
    return completed.stdout


def chart_metadata(chart_dir: Path) -> tuple[str, str]:
    metadata: dict[str, str] = {}
    for line in (chart_dir / "Chart.yaml").read_text(encoding="utf-8").splitlines():
        if ":" not in line or line.startswith((" ", "\t", "#")):
            continue
        key, value = line.split(":", 1)
        if key in {"name", "version"}:
            metadata[key] = value.strip().strip("\"'")
    return metadata["name"], metadata["version"]


def changed_chart_names(base_ref: str, head_ref: str = "HEAD", *, root: Path = ROOT) -> set[str]:
    changed_files = run(["git", "diff", "--name-only", f"{base_ref}...{head_ref}"], cwd=root)
    charts: set[str] = set()
    for changed_file in changed_files.splitlines():
        parts = changed_file.split("/", 2)
        if len(parts) < 2 or parts[0] != "charts":
            continue
        chart_name = parts[1]
        if (root / "charts" / chart_name / "Chart.yaml").is_file():
            charts.add(chart_name)
    return charts


def tag_exists(tag_name: str, *, root: Path = ROOT) -> bool:
    completed = subprocess.run(
        ["git", "rev-parse", "--verify", "--quiet", f"refs/tags/{tag_name}"],
        cwd=root,
        check=False,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )
    return completed.returncode == 0


def validate_changed_chart_versions(
    *,
    base_ref: str,
    head_ref: str = "HEAD",
    root: Path = ROOT,
) -> list[str]:
    failures: list[str] = []
    charts = changed_chart_names(base_ref, head_ref, root=root)
    if not charts:
        print("no changed charts require release version validation")
        return failures

    for chart_name in sorted(charts):
        release_name, version = chart_metadata(root / "charts" / chart_name)
        tag_name = f"{release_name}-{version}"
        if tag_exists(tag_name, root=root):
            failures.append(
                f"{chart_name}: release tag {tag_name} already exists; "
                f"bump charts/{chart_name}/Chart.yaml version"
            )
        else:
            print(f"{chart_name}: release tag {tag_name} is available")

    return failures


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--base", required=True, help="Base git ref for changed-file detection")
    parser.add_argument("--head", default="HEAD", help="Head git ref for changed-file detection")
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    try:
        failures = validate_changed_chart_versions(base_ref=args.base, head_ref=args.head)
    except Exception as exc:  # noqa: BLE001 - normalize CI errors at the script seam.
        print(f"::error::{exc}", file=sys.stderr)
        return 1

    if failures:
        for failure in failures:
            print(f"::error::{failure}", file=sys.stderr)
        return 1

    print("chart release versions ok")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: .github/scripts/validate_chart_release_versions.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Validate that changed charts declare an unreleased chart version.
#
# Chart releaser publishes releases as <chart-name>-<chart-version>. If a chart
# changes without a version bump, the release workflow fails after merge with a
# GitHub "tag already exists" error. This script moves that invariant into CI.

require "open3"
require "yaml"
require "set"
require "optparse"

ROOT = File.expand_path("../..", __dir__)
CHARTS_DIR = File.join(ROOT, "charts")

options = {
  base_ref: ENV["CHART_VERSION_BASE_REF"],
  head_ref: ENV.fetch("CHART_VERSION_HEAD_REF", "HEAD")
}

OptionParser.new do |parser|
  parser.on("--base REF", "Base git ref for changed-file detection") { |value| options[:base_ref] = value }
  parser.on("--head REF", "Head git ref for changed-file detection") { |value| options[:head_ref] = value }
end.parse!

def run!(*command)
  stdout, stderr, status = Open3.capture3(*command, chdir: ROOT)
  unless status.success?
    warn stderr
    raise "#{command.join(' ')} failed"
  end
  stdout
end

def chart_metadata(chart_name)
  path = File.join(CHARTS_DIR, chart_name, "Chart.yaml")
  metadata = YAML.safe_load(File.read(path))
  name = metadata.fetch("name")
  version = metadata.fetch("version")
  [name, version]
end

def changed_chart_names(base_ref, head_ref)
  raise "base ref is required for chart release version validation" if base_ref.to_s.empty?

  changed_files = run!("git", "diff", "--name-only", "#{base_ref}...#{head_ref}").lines.map(&:strip)
  changed_files.each_with_object(Set.new) do |path, charts|
    next unless path.start_with?("charts/")

    _charts_dir, chart_name, = path.split("/", 3)
    next if chart_name.to_s.empty?
    next unless File.file?(File.join(CHARTS_DIR, chart_name, "Chart.yaml"))

    charts << chart_name
  end
end

def tag_exists?(tag_name)
  system("git", "rev-parse", "--verify", "--quiet", "refs/tags/#{tag_name}", chdir: ROOT, out: File::NULL)
end

begin
  charts = changed_chart_names(options.fetch(:base_ref), options.fetch(:head_ref))

  if charts.empty?
    puts "no changed charts require release version validation"
    exit 0
  end

  failures = []
  charts.sort.each do |chart_name|
    release_name, version = chart_metadata(chart_name)
    tag_name = "#{release_name}-#{version}"
    if tag_exists?(tag_name)
      failures << "#{chart_name}: release tag #{tag_name} already exists; bump charts/#{chart_name}/Chart.yaml version"
    else
      puts "#{chart_name}: release tag #{tag_name} is available"
    end
  end

  unless failures.empty?
    failures.each { |failure| warn "::error::#{failure}" }
    exit 1
  end

  puts "chart release versions ok"
rescue StandardError => e
  warn "::error::#{e.message}"
  exit 1
end
